File: FallbackResponse.java
package javaapplication2;

import java.util.function.Consumer;
import java.util.function.Function;

public final class FallbackResponse<R, E extends Exception> {

    private final R result;
    private final E error;

    public FallbackResponse(R result, E error) {
        this.result = result;
        this.error = error;
    }

    public <T> FallbackResponse<T, E> call(Function<R, T> func) {
        try {
            var response = func.apply(this.result);
            return new FallbackResponse(response, null);
        } catch (Exception ex) {
            return new FallbackResponse(null, error);
        }
    }

    public FallbackResponse<R, E> whenErrorDo(Consumer<E> act) {
        if (this.error != null) {
            act.accept(this.error);
        }
        return this;
    }

    public <T extends Exception> FallbackResponse<R, E> whenErrorDo(Class<T> errorType, Consumer<T> act) {
        if (this.error != null && errorType.isAssignableFrom(this.error.getClass())) {
            act.accept((T)this.error);
        }
        return this;
    }

    public FallbackResponse<R, E> whenErrorDoAndThrow(Consumer<E> act) throws E {
        if (this.error != null) {
            act.accept(this.error);
            throw this.error;
        }
        return this;
    }

    public <T extends Exception> FallbackResponse<R, E> whenErrorDoAndThrow(Class<T> errorType, Consumer<T> act) throws E {
        if (this.error != null && errorType.isAssignableFrom(this.error.getClass())) {
            act.accept((T) this.error);
            throw this.error;
        }
        return this;
    }

    public FallbackResponse<R, E> whenErrorThrow() throws E {
        if (this.error != null) {
            throw this.error;
        }
        return this;
    }

    public <T extends Exception> FallbackResponse<R, E> whenErrorThrow(Class<T> errorType) throws E {
        if (this.error != null && errorType.isAssignableFrom(this.error.getClass())) {
            throw this.error;
        }
        return this;
    }

    public R getResult() {
        return result;
    }

    public R getResult(R def) {
        return result == null ? def : result;
    }

    public E getError() {
        return error;
    }

}

File: FallbackResponse.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using static System.Runtime.InteropServices.JavaScript.JSType;

namespace ConsoleApp2
{
    public sealed class FallbackResponse<T>
    {
        private readonly T? result;
        private readonly Exception? error;

        public FallbackResponse(T? result, Exception? error)
        {
            this.result = result;
            this.error = error;
        }

        public FallbackResponse<T2> Call<T2>(Func<T?, T2> action)
        {
            try
            {
                var resp = action(result);
                return new FallbackResponse<T2>(resp, error);
            }
            catch (Exception ex)
            {
                return new FallbackResponse<T2>(default, ex);
            }
        }

        public Exception? Error() => error;
        public T Result(T def) => result == null ? def : result;
        public T? Result() => result;

        public FallbackResponse<T> WhenErrorDo(Action<Exception> act)
        {
            if (error != null)
                act(error);
            return this;
        }

        public FallbackResponse<T> WhenErrorDo<E>(Action<E> act) where E : Exception
        {
            if (error != null && error is E)
                act((E)error);
            return this;
        }

        public FallbackResponse<T> WhenErrorDoAndThrow(Action<Exception?> act)
        {
            if (error != null)
            {
                act(error);
                throw error;
            }
            return this;
        }

        public FallbackResponse<T> WhenErrorDoAndThrow<E>(Action<E> act) where E : Exception
        {
            if (error != null && error is E)
            {
                act((E)error);
                throw error;
            }
            return this;
        }

        public FallbackResponse<T> WhenErrorThrow()
        {
            if (error != null)
                throw error;
            return this;
        }

        public FallbackResponse<T> WhenErrorThrow<E>()
        {
            if (error != null && error is E)
                throw error;
            return this;
        }

    }
}
